include/camp/device.hpp

```diff
@@ -0,0 +1,273 @@
+	  bool check() const override { return m_modelVal.check(); }
+	  void wait() const override { m_modelVal.wait(); }
+	  T get() { return m_modelVal; }
+	private:
+	  T m_modelVal;
+      };
+
+      std::shared_ptr<EventInterface> m_value;
+  };
+
+  class Cuda 
+  {
+    static cudaStream_t get_a_stream(int num)
+    {
+      static cudaStream_t streams[16] = {};
+      static int previous = 0;
+
+      static std::once_flag m_onceFlag;
+      static std::mutex m_mtx;
+
+      std::call_once(m_onceFlag,
+	[] {
+	  if (streams[0] == nullptr) {
+	    for (auto &s : streams) {
+	      cudaStreamCreate(&s);
+	    }
+	  }
+	});
+
+      if (num < 0) {
+	m_mtx.lock();
+        previous = (previous + 1) % 16;
+	m_mtx.unlock();
+        return streams[previous];
+      }
+
+      return streams[num % 16];
+    }
+
+  public:
+    Cuda(int device = 0, int group = -1) : stream(get_a_stream(group)) {}
+
+    // Methods
+    Platform get_platform() { return Platform::cuda; }
+    static Cuda &get_default()
+    {
+      static Cuda h;
+      return h;
+    }
+    CudaEvent get_event() { 
+      return CudaEvent(get_stream());
+    }
+    Event get_event_erased() {
+      return Event{CudaEvent(get_stream())};
+    }
+    void wait() { cudaStreamSynchronize(stream); }
+    void wait_on(Event *e) { e->wait(); }
+
+    // Memory
+    template <typename T>
+    T *allocate(size_t size)
+    {
+      T *ret = nullptr;
+      cudaMallocManaged(&ret, sizeof(T) * size);
+      return ret;
+    }
+    void *calloc(size_t size)
+    {
+      void *p = allocate<char>(size);
+      this->memset(p, 0, size);
+      return p;
+    }
+    void free(void *p) { cudaFree(p); }
+    void memcpy(void *dst, const void *src, size_t size)
+    {
+      cudaMemcpyAsync(dst, src, size, cudaMemcpyDefault, stream);
+    }
+    void memset(void *p, int val, size_t size)
+    {
+      cudaMemsetAsync(p, val, size, stream);
+    }
+
+    cudaStream_t get_stream() { return stream; }
+
+  private:
+    cudaStream_t stream;
+  };
+
+  class Host
+  {
+  public:
+    Host(int device = 0, int group = -1) {}
+
+    // Methods
+    Platform get_platform() { return Platform::host; }
+    static Host &get_default()
+    {
+      static Host h;
+      return h;
+    }
+    HostEvent get_event() { return HostEvent(); }
+    Event get_event_erased() {
+      Event e{HostEvent()};
+      return e;
+    }
+    void wait() {}
+    void wait_on(Event *e) { }
+
+    // Memory
+    template <typename T>
+    T *allocate(size_t n)
+    {
+      return (T*)malloc(sizeof(T) * n);
+    }
+    void *calloc(size_t size)
+    {
+      void *p = allocate<char>(size);
+      this->memset(p, 0, size);
+      return p;
+    }
+    void free(void *p) { free(p); }
+    void memcpy(void *dst, const void *src, size_t size)
+    {
+      memcpy(dst, src, size);
+    }
+    void memset(void *p, int val, size_t size) { std::memset(p, val, size); }
+  };
+
+  class Context
```

Comment:
We've changed the name of the underlying base class from "Device" to "Context", do we want to rename the namespace form "devices" to something else?  @trws

Reply:
Sure, how about "resource" or "resources"?  @davidbeckingsale, thoughts?

Reply:
Do we need to determine between an execution resource and another kind of resource?

Reply:
Perhaps?  Each of these has both compute and memory associations, so it's kinda both.